src/shared/components/enhancedReactTable/EnhancedReactTable.tsx

```diff
@@ -106,33 +93,42 @@ export default class EnhancedReactTable<T> extends React.Component<IEnhancedReac
 
     private columnVisibilitySelector = (state:IEnhancedReactTableState) => state.columnVisibility;
 
+    // visible columns depend on both column definitions and visibility
     private visibleColsSelector = createSelector([this.columnsSelector, this.columnVisibilitySelector],
         (columns:IColumnDefMap, columnVisibility:IColumnVisibilityState) => this.resolveVisible(columns, columnVisibility));
 
-    private sortableColsSelector = createSelector(this.visibleColsSelector,
-        (visibleCols:IColumnDefMap)=>this.resolveSortable(visibleCols));
-
-    private filterableColsSelector = createSelector(this.visibleColsSelector,
-        (visibleCols:IColumnDefMap)=>this.resolveFilterable(visibleCols));
-
-    private sortedColsSelector = createSelector(this.visibleColsSelector,
+    // sorted list of visible columns are required to calculate table header
+    private sortedVisibleColsSelector = createSelector(this.visibleColsSelector,
         (visibleCols:IColumnDefMap)=>this.resolveOrder(visibleCols));
 
-    private columnVisibilityArraySelector = createSelector(this.columnVisibilitySelector,
-        (columnVisibility:IColumnVisibilityState) => this.resolveColumnVisibility(this.colNameToId, this.sortedColumns, columnVisibility));
+    // sorted columns depend on column definition, but not on the visibility
+    private sortedColsSelector = createSelector(this.columnsSelector,
+        (columns:IColumnDefMap)=>this.resolveOrder(columns));
 
-    private headersSelector = createSelector(this.sortedColsSelector,
+    private columnVisibilityArraySelector = createSelector([this.sortedColsSelector, this.columnVisibilitySelector],
+         (sortedColumns:Array<IEnhancedReactTableColumnDef>, columnVisibility:IColumnVisibilityState) =>
+             this.resolveColumnVisibility(this.colNameToId, sortedColumns, columnVisibility));
+
+    // we need to calculate the header values every time the visibility or column definitions change
+    private headersSelector = createSelector(this.sortedVisibleColsSelector,
         (sortedCols:Array<IEnhancedReactTableColumnDef>) => this.generateHeaders(sortedCols));
 
+    // no need to calculate column render values every time the visibility changes,
+    // we only need to do it when the column definitions change
+    // reactable is clever enough to not render the column if its header is missing
     private rowsSelector = createSelector([this.sortedColsSelector, this.rawDataSelector],
         (sortedCols:Array<IEnhancedReactTableColumnDef>, rawData:Array<T>) => this.generateRows(sortedCols, rawData));
 
     constructor(props:IEnhancedReactTableProps<T>)
     {
         super(props);
 
+        this.colNameToId = this.mapColNameToId(props.columns);
+
         this.state = {
             columnVisibility: EnhancedReactTable.resolveVisibility(props.columns, props.rawData),
+            columnSort: this.resolveSortable(this.props.columns || {}),
```

Comment:
this name doesn't make sense to me/is sort of vague. It's not quite a verb, not quite a noun. I think it should be sortableColumns or something. same with columnFilter